snapshot_creator.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 the v8go contributors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package v8go
+
+// #include <stdlib.h>
+// #include "v8go.h"
+import "C"
+import (
+	"errors"
+	"unsafe"
+)
+
+type FunctionCodeHandling int
+
+//  Clear - does not keep any compiled data prior to serialization/deserialization/verify pass
+//  Keep - keeps any compiled data prior to serialization/deserialization/verify pass
+const (
+	FunctionCodeHandlingClear FunctionCodeHandling = iota
+	FunctionCodeHandlingKeep
+)
+
+// StartupData stores the snapshot blob data
+type StartupData struct {
+	data     []byte
+	raw_size C.int
+}
+
+// SnapshotCreator allows creating snapshot.
+type SnapshotCreator struct {
+	ptr                 C.SnapshotCreatorPtr
+	iso                 *Isolate
+	defaultContextAdded bool
+}
+
+// NewSnapshotCreator creates a new snapshot creator.
+func NewSnapshotCreator() *SnapshotCreator {
+	v8once.Do(func() {
+		C.Init()
+	})
+
+	rtn := C.NewSnapshotCreator()
+
+	return &SnapshotCreator{
+		ptr:                 rtn.creator,
+		iso:                 &Isolate{ptr: rtn.iso},
+		defaultContextAdded: false,
+	}
+}
+
+// GetIsolate returns the Isolate associated with the SnapshotCreator.
+// This Isolate must be used to create the contexts that later will be used to create the snapshot blob.
+func (s *SnapshotCreator) GetIsolate() (*Isolate, error) {
+	if s.ptr == nil {
+		return nil, errors.New("v8go: Cannot get Isolate after creating the blob")
+	}
+
+	return s.iso, nil
+}
+
+// SetDefaultContext set the default context to be included in the snapshot blob.
+func (s *SnapshotCreator) SetDefaultContext(ctx *Context) error {
+	if s.defaultContextAdded {
+		return errors.New("v8go: Cannot set multiple default context for snapshot creator")
+	}
+
+	C.SetDefaultContext(s.ptr, ctx.ptr)
+	s.defaultContextAdded = true
+	ctx.ptr = nil
+
+	return nil
+}
+
+// AddContext add additional context to be included in the snapshot blob.
+// Returns the index of the context in the snapshot blob, that later can be use to call v8go.NewContextFromSnapshot.
+func (s *SnapshotCreator) AddContext(ctx *Context) (int, error) {
+	if s.ptr == nil {
+		return 0, errors.New("v8go: Cannot add context to snapshot creator after creating the blob")
+	}
+
+	index := C.AddContext(s.ptr, ctx.ptr)
+	ctx.ptr = nil
+
+	return int(index), nil
+}
+
+// Create creates a snapshot data blob.
+// The snapshot creator instance is unsable after creating the snapshot data blob.
+func (s *SnapshotCreator) Create(functionCode FunctionCodeHandling) (*StartupData, error) {
```

Comment:
StartupData is basically just a wrapper around a slice.  It may as well be returned without the extra pointer indirection, since the error would indicate whether or not it is an error.  Since V8 doesn't actually expose an error message, `ok` could be used for the error and the `s.ptr == nil` and `!s.defaultContextAdded` cases could result in a panic.